holidays/countries/japan.py

```diff
@@ -51,135 +51,160 @@ def _populate(self, year):
-            )
-
-        # Autumnal Equinox Day.
         epoch = Sun.get_equinox_solstice(year, target="autumn")
         equinox = map(int, Epoch(epoch).get_full_date())
         adjusted_date = (
             datetime(*equinox, tzinfo=timezone.utc)
             .astimezone(ZoneInfo("Asia/Tokyo"))
             .date()
         )
-        self._add_holiday(tr("秋分の日"), adjusted_date)
-
-        # Health and Sports Day.
-        if 1966 <= year <= 1999:
-            self._add_holiday(tr("体育の日"), OCT, 10)
-        elif 2000 <= year <= 2019:
-            self._add_holiday(
-                tr("体育の日"), _get_nth_weekday_of_month(2, MON, OCT, year)
+        # Autumnal Equinox Day.
+        observed_dates.add(self._add_holiday(tr("秋分の日"), adjusted_date))
+
+        # Physical Education and Sports Day.
+        if year >= 1966:
+            name = (
+                # Sports Day.
+                tr("スポーツの日")
+                if year >= 2020
+                # Physical Education Day.
+                else tr("体育の日")
             )
-        elif year == 2020:
-            self._add_holiday(tr("スポーツの日"), JUL, 24)
-        elif year == 2021:
-            self._add_holiday(tr("スポーツの日"), JUL, 23)
-        elif 2022 <= year:
-            self._add_holiday(
-                tr("スポーツの日"), _get_nth_weekday_of_month(2, MON, OCT, year)
+            dates = {
+                2020: date(2020, JUL, 24),
+                2021: date(2021, JUL, 23),
+            }
+            dt = dates.get(
+                year,
+                _get_nth_weekday_of_month(2, MON, OCT, year)
+                if year >= 2000
+                else date(year, OCT, 10),
             )
+            observed_dates.add(self._add_holiday(name, dt))
 
         # Culture Day.
-        self._add_holiday(tr("文化の日"), NOV, 3)
+        observed_dates.add(self._add_holiday(tr("文化の日"), NOV, 3))
 
-        # Labour Thanksgiving Day.
-        self._add_holiday(tr("勤労感謝の日"), NOV, 23)
+        # Labor Thanksgiving Day.
+        observed_dates.add(self._add_holiday(tr("勤労感謝の日"), NOV, 23))
 
         # Regarding the Emperor of Heisei.
         if 1989 <= year <= 2018:
-            self._add_holiday(tr("天皇誕生日"), DEC, 23)
+            observed_dates.add(self._add_holiday(tr("天皇誕生日"), DEC, 23))
 
         if self.observed:
+            for month, day, _ in self.special_holidays.get(year, ()):
+                observed_dates.add(date(year, month, day))
+
             # When a national holiday falls on Sunday, next working day
             # shall become a public holiday (振替休日) - substitute holidays.
-            for dt in list(self.keys()):
-                if dt.year == year and self._is_sunday(dt):
-                    hol_date = dt + td(days=+1)
-                    while hol_date in self:
-                        hol_date += td(days=+1)
-                    self._add_holiday(tr("振替休日"), hol_date)
+            for dt in set(observed_dates):
```

Comment:
Looks odd taking into account that observed_dates is a `set` already.
Maybe
```suggestion
            for dt in observed_dates.copy():
```